sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/Identity.java

```diff
@@ -492,89 +518,114 @@ private IAccount retrieveAccount(String id) {
+            future.complete(new SignInResult(null, new IllegalStateException("signIn is called while it's not configured or not in the foreground.")));
         }
     }
 
-    private synchronized void silentSignIn(@NonNull IAccount account) {
+    private synchronized void silentSignIn(final DefaultAppCenterFuture<SignInResult> future, @NonNull IAccount account, final boolean withUIFallback) {
         AppCenterLog.info(LOG_TAG, "Sign in silently in the background.");
         mAuthenticationClient.acquireTokenSilentAsync(new String[]{mIdentityScope}, account, null, true, new AuthenticationCallback() {
 
             @Override
             public void onSuccess(IAuthenticationResult authenticationResult) {
-                handleSignInSuccess(authenticationResult);
+                handleSignInSuccess(future, authenticationResult);
             }
 
             @Override
             public void onError(MsalException exception) {
-                if (exception instanceof MsalUiRequiredException) {
+                if (withUIFallback && exception instanceof MsalUiRequiredException) {
                     AppCenterLog.info(LOG_TAG, "No token in cache, proceed with interactive sign-in experience.");
                     postOnUiThread(new Runnable() {
 
                         @Override
                         public void run() {
-                            signInInteractively();
+                            signInInteractively(future);
                         }
                     });
                 } else {
-                    handleSignInError(exception);
+                    handleSignInError(future, exception);
                 }
             }
 
             @Override
             public void onCancel() {
-                handleSignInCancellation();
+                handleSignInCancellation(future);
             }
         });
     }
 
-    private void handleSignInSuccess(final IAuthenticationResult authenticationResult) {
+    private synchronized void refreshToken(String homeAccountId, boolean networkConnected) {
+        if (mLastSignInFuture != null && !mLastSignInFuture.isDone()) {
+            AppCenterLog.debug(LOG_TAG, "Failed to refresh token: sign-in already in progress.");
+            return;
+        }
+        if (mLastRefreshFuture != null && !mLastRefreshFuture.isDone()) {
+            AppCenterLog.verbose(LOG_TAG, "Token refresh already in progress. Skip this refresh request.");
+            return;
+        }
+        if (!networkConnected) {
+            mHomeAccountIdToRefresh = homeAccountId;
+            AppCenterLog.debug(LOG_TAG, "Network not connected. The token will be refreshed after coming back online.");
+            return;
+        }
+        final DefaultAppCenterFuture<SignInResult> future = new DefaultAppCenterFuture<>();
+        mLastRefreshFuture = future;
+        IAccount account = retrieveAccount(homeAccountId);
```

Comment:
Before lollipop, onNetworkState is from UI thread.
This method accesses disk.
Need to post() from onNetworkStateUpdated and review synchronized again.